File: palindromeLinkedList.cpp
//time complexity: O(n)
//space complexity:O(1)
//approach: reversing the later half of the list and comparing it with the first half.
//solved on leetcode? yes
//problem faced? no
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if(head==NULL || head->next==NULL)
            return true;
        ListNode* slow=head;
        ListNode* fast=head;
        while(fast!=NULL && fast->next!=NULL)
        {
            slow=slow->next;
            fast=fast->next->next;
        }
        ListNode* prev=NULL;
        ListNode* curr=slow;
        ListNode* next;
        while(curr!=NULL)
        {
            next=curr->next;
            curr->next=prev;
            prev=curr;
            curr=next;
        }
        ListNode* p=head;
        while(prev!=NULL)
        {
            if(p->val!=prev->val)
                return false;
            p=p->next;
            prev=prev->next;
        }
        return true;
    }
};

File: cc_4/palindrome_Linked_List.java
package cc_4;

public class palindrome_Linked_List {
    //TC: O(n)
    //SC: O(1)
    public boolean isPalindrome(ListNode head) {
        if(head==null) return true;
        ListNode fast = head;
        ListNode slow = head;
        while(fast!=null && fast.next!=null){
            fast=fast.next.next;
            slow=slow.next;
        }
        int result = compare(slow, head);
        if(result==-1) return false;
        return true;
    }
    private int compare(ListNode slow, ListNode fast){
        ListNode reversedLL = reverseLL(slow);
        slow=fast;
        while(reversedLL!=null && slow!=null){
            if(slow.val!=reversedLL.val){
                return -1;
            }
            slow = slow.next;
            reversedLL = reversedLL.next;
        }
        return 1;
    }
    private ListNode reverseLL(ListNode head){
        ListNode curr = head;
        ListNode prev = null;
        while(curr!=null)
        {
            ListNode nextnode = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextnode;
        }
        return prev;
    }
}
